src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperUseageDescriptorFactory.cs

```diff
@@ -0,0 +1,144 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+#if !DNXCORE50 // Cannot accurately resolve the location of the documentation XML file in coreclr.
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.IO;
+using System.Linq;
+using System.Reflection;
+using System.Xml.Linq;
+using Microsoft.AspNet.Razor.TagHelpers;
+using Microsoft.Framework.Internal;
+
+namespace Microsoft.AspNet.Razor.Runtime.TagHelpers
+{
+    public static class TagHelperUseageDescriptorFactory
+    {
+        private static readonly Dictionary<string, IEnumerable<XElement>> XMLFileMembersCache =
+            new Dictionary<string, IEnumerable<XElement>>(StringComparer.Ordinal);
+        private static readonly char[] InvalidPathCharacters = Path.GetInvalidPathChars();
+
+        public static TagHelperUseageDescriptor CreateDescriptor([NotNull] TypeInfo typeInfo)
+        {
+            var lookupName = $"T:{typeInfo.FullName}";
+            return CreateDescriptorCore(typeInfo.Assembly, lookupName);
+        }
+
+        public static TagHelperUseageDescriptor CreateDescriptor([NotNull] PropertyInfo propertyInfo)
+        {
+            var declaringTypeInfo = propertyInfo.DeclaringType.GetTypeInfo();
+            var lookupName = $"P:{declaringTypeInfo.FullName}.{propertyInfo.Name}";
+            return CreateDescriptorCore(declaringTypeInfo.Assembly, lookupName);
+        }
+
+        private static TagHelperUseageDescriptor CreateDescriptorCore(Assembly typeAssembly, string lookupName)
+        {
+            var typeAssemblyLocation = typeAssembly.Location;
+
+            if (string.IsNullOrWhiteSpace(typeAssemblyLocation) && !string.IsNullOrWhiteSpace(typeAssembly.CodeBase))
+            {
+                var uri = new UriBuilder(typeAssembly.CodeBase);
+
+                // Normalize the path to a UNC path. This will remove things like file:// from start of the uri.Path.
+                typeAssemblyLocation = Uri.UnescapeDataString(uri.Path);
+
+                // Still couldn't resolve a valid typeAssemblyLocation.
+                if (string.IsNullOrWhiteSpace(typeAssemblyLocation))
+                {
+                    return null;
+                }
+            }
+
+            IEnumerable<XElement> members = null;
+            if (!XMLFileMembersCache.TryGetValue(typeAssemblyLocation, out members))
+            {
+                var xmlDocumentationFile = GetXmlDocumentationFile(typeAssemblyLocation);
+
+                // We only want to process the file if it exists. In the case it doesn't, a null value will be added
+                // to the cache to not constantly look for new XML files.
+                if (xmlDocumentationFile != null)
+                {
+                    var xmlDocumentation = XDocument.Load(xmlDocumentationFile.FullName);
```

Comment:
Suggestion by @rynowak is to break out the parsing code of the XML file into a stand alone class in-case we want to utilize it in other areas down the line.